### ISSUE TITLE
Finetune video predictor on custom data

### ISSUE THREAD

**Issue by ahmadsadeed** [facebookresearch/segment-anything-2#305]
I was able to finetune SAM on custom image segmentation dataset which trained the mask decoder and got very high IoU. Then I used the same model weights to test video predictor but didn't see improvement. I assume it has to do with memory encoder. How can I finetune the whole model? Didn't want to reinvent the wheel if someone has done it.

**Comment by kyzira** [comment]
I have the same issue, did you find a solution?

**Comment by tutu941008** [comment]
> 我能够在自定义图像分割数据集上微调 SAM，该数据集训练了掩码解码器并获得非常高的 IoU。然后，我使用相同的模型权重来测试视频预测器，但没有看到改进。我认为这与 memory encoder 有关。如何微调整个模型？如果有人做了，我不想重新发明轮子。

After fine-tuning on my own dataset, I obtained a new model weights file (`.pt`) and the corresponding `.yaml` configuration file used during training. Now that I have these files, how can I use the new weights file to predict my images?

**Comment by kyzira** [comment]
how did you train your model? If you trained it like the example in the training folder, you can just swap out the paths of the models.
For example instead of:

```python
from sam2.build_sam import build_sam2_video_predictor

sam2_checkpoint = "../checkpoints/sam2.1_hiera_large.pt"
model_cfg = "configs/sam2.1/sam2.1_hiera_l.yaml"

predictor = build_sam2_video_predictor(model_cfg, sam2_checkpoint, device=device)
```

you can use:

```python
from sam2.build_sam import build_sam2_video_predictor

sam2_checkpoint = "C:\Users\trained_sam_model_path\trained_checkpoint_of_large_model.pt"
model_cfg = "configs/sam2.1/sam2.1_hiera_l.yaml"

predictor = build_sam2_video_predictor(model_cfg, sam2_checkpoint, device=device)
```

You will need to match the model config to the model size you trained. (eg. when you trained the base plus model, use: `configs\sam2.1\sam2.1_hiera_b+.yaml`)

For me it only worked using the configs from the "configs/sam2.1/" directory and not the one used to train the model.

**Comment by sheldonsebastian** [comment]
Hey [kyzira](https://github.com/kyzira), when you train the model for video prediction, did the training loss go down? I'm having very high dice loss when I used a combination of SA-V dataset and MOSE subset data.

**Comment by sucessfullys** [comment]
> > 我能够在自定义图像分割数据集上参数SAM，该数据集训练了掩码解码器并获得了非常高的IoU。然后，我使用相同的模型权重来测试视频预测器，但没有看到任何改进。我认为这与内存编码器有关。如何影响整个模型？如果有人做到了，我不想重新发明轮子。
> 
> 在自己的数据集上进行微调后，我获得了一个新的模型权重文件（`.pt`）以及训练时使用的相应`.yaml`配置文件。现在我有了这些文件，我该如何使用新的权重文件来预测我的图像？

你好，你是用什么数据集呢

**Comment by 2nzi** [comment]
> For me it only worked using the configs from the "configs/sam2.1/" directory and not the one used to train the model.

@kyzira Did you get good results when fine-tuning your model? Do you have any new insights or methods for fine-tuning on videos and improving tracking for a specific task (i am in sport tracking)?

**Comment by kyzira** [comment]
@2nzi I still use it the way i described above. You have to convert your labelled dataset into the MOSE format but i get pretty good results. After labeling around 200-300 video snippets, i convert my labels again into the MOSE format and retrain sam2 and everytime the predictions get better.

For example if i label cracks on concrete, without training it is very hard to label them with less than 5 inputs, now i just click once and its labelled pretty easily

**Comment by 2nzi** [comment]
> [@2nzi](https://github.com/2nzi) I still use it the way i described above. You have to convert your labelled dataset into the MOSE format but i get pretty good results. After labeling around 200-300 video snippets, i convert my labels again into the MOSE format and retrain sam2 and everytime the predictions get better.

> For example if i label cracks on concrete, without training it is very hard to label them with less than 5 inputs, now i just click once and its labelled pretty easily

@kyzira Thanks for your answer!

So, do you use it to segment images or videos?
What tool do you use to label/segment your images or video frames?
Do you happen to have an example script or notebook showing how to fine-tune SAM with the MOSE format?


(I’m currently an intern in a research lab working on segmenting sports players. I’m learning how SAM works and how to adapt or improve it for my use case.)

**Comment by kyzira** [comment]
I was also an intern, i created a labelling tool with tkinter in python, where you click on the image, which is displayed in a canvas. Depending on Left or right click a positive or negativ point is added.
I cut up the video to 1 frame per second and put them into one folder. this is the folder i show sam2's video tracking.
From my experience it still can track objects well enough with 1 frame per second.


below i created a md file, for finetuning a sam2 model:


How to train SAM2 on custom dataset

Here i will show you how to train sam2 on your labeled custom dataset
1. Setup your Machine

Training only works on linux machines, so if on windows [please install wsl2 with ubuntu](https://learn.microsoft.com/en-us/windows/wsl/install).
1. CUDA installation

Now in the terminal we need cuda first

wget https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2404/x86_64/cuda-keyring_1.1-1_all.deb
sudo apt install ./cuda-keyring_1.1-1_all.deb
sudo apt update
sudo apt install cuda-toolkit

Set Cuda in the PATH:

nano /home/$USER/.bashrc

Add at the end of the file

export CUDA_HOME=/usr/local/cuda
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/local/cuda/lib64:/usr/local/cuda/extras/CUPTI/lib64
export PATH=$PATH:$CUDA_HOME/bin

Then terminal:

source .bashrc

Now run

nvcc --version

2. Anaconda Installation

now download anaconda from [here](https://www.anaconda.com/download/success) and put it into your user folder in wsl.

then to run the shell just type in the name of the file. For example:

Anaconda-latest-Linux-x86_64.sh

Scroll to the bottom of the EULA with enter and type yes and hit enter twice to start the installation. At the end they will ask if you want to use conda as your standard terminal, just type yes and enter.

Restart your terminal
3. Pytorch

Now lets install pytorch. Select Stable, Linux, Conda, Python and the latest Cuda version, copy the command at the bottom and run it.

type y when ask if you want to proceed.
4. SAM2

[Here is its Github Page](https://github.com/facebookresearch/sam2)

To install it and everything it needs, run the following commands:

git clone https://github.com/facebookresearch/sam2.git && cd sam2
pip install -e .
pip install -e ".[notebooks]"
cd checkpoints
./download_ckpts.sh
cd ..
pip install -e ".[dev]"

2. Prepare your Dataset

Your Dataset has to be in a specific format. We will use the MOSE Dataset format
MOSE Dataset format

This format looks like this:


├── BinaryMasks
│ │ 
│ ├── <video_name_1>
│ │ ├── 00000.png
│ │ ├── 00001.png
│ │ └── ...
│ │ 
│ ├── <video_name_2>
│ │ ├── 00000.png
│ │ ├── 00001.png
│ │ └── ...
│ │ 
│ ├── <video_name_...>
│ 
└── JPEGImages
  │ 
  ├── <video_name_1>
  │ ├── 00000.jpg
  │ ├── 00001.jpg
  │ └── ...
  │ 
  ├── <video_name_2>
  │ ├── 00000.jpg
  │ ├── 00001.jpg
  │ └── ...
  │ 
  └── <video_name_...>

The Masks should be binary images with the mask having values of 1. Make sure, that the folders do not include a video filetype in their names, e.g. video1 instead of video1.mpg
Convert Dataset

If you have used this tool to label your data, just run the prepare_mose_format_dataset.py file. First put in the directroy of the labeled results and the output directory. Different object masks are drawn as different 255 value colors. technically it should work, because it is done so in the mose dataset, but somehow it doesnt. Set the use_single_color_mode variable to True, when it doesnt work.
Put Converted Dataset into your user folder

After converting dont forget to copy your dataset to the ubuntu user folder.
3. Prepare train config

If you have your data ready, now we can prepare the yaml file we are going to run. Open the file under: sam2/sam2/configs/sam2.1_training/sam2.1_hiera_b+_MOSE_finetune.yaml

On the top of the file you can now input the paths to your dataset images and masks. Change gpus per node to 1 lower at the bottom and also edit the save_freq variable to save your checkpoint every x epochs. You can also set the seed value to another value, that should change the outcome a bit, when retraining the same model.
Quirks

When setting the Parameters i encountered some errors, which came from changing the values:

    frame num: Set frame num to 1
    num of Epochs: Dont go under 40, epoch sizes of 10 or 20 didnt work for me.
    file_list_txt: set it to null

if you get error, you can try setting feat_sizes to [32, 32]
4. Run Training
Run Training

Finally you can run the training with: Make sure to be in the sam2 root directory

python training/train.py \
    -c configs/sam2.1_training/sam2.1_hiera_b+_MOSE_finetune.yaml \
    --use-cluster 0 \
    --num-gpus 1

Tensorboard

You can watch live statistics if you open an other terminal and then run:

tensorboard --logdir sam2/sam2_logs/configs/sam2.1_training/sam2.1_hiera_b+_MOSE_finetune.yaml/tensorboard

Now you can ctrl + leftclick the ip adress which pops up and you can see live graphs of the stats.
5. Use your Checkpoint

When completed, you can use your new checkpoint, instead of the one from meta. it will be saved under sam2/sam2_logs/configs/sam2.1_training/sam2.1_hiera_b+_MOSE_finetune.yaml/checkpoints.

As the model config you have to use sam2\sam2\configs\sam2.1\sam2.1_hiera_b+.yaml

**Comment by 2nzi** [comment]
@kyzira thanks for all the advice, I’ll give it a try!
I have another question:
If you have multiple objects in the same video, did you create a separate video and mask for each object, or did you include all objects in a single mask for each frame?
In your example, did you sometimes have multiple objects annotated at once in the same frame?

**Comment by kyzira** [comment]
@2nzi When tracking multiple Objects in one video i load the ~20-50 frames from the video into sam2. (I use one frame per second)
With that now it is possible to input all points you want to set and track multiple objects at once but i used a different way.
For optimization i let the User choose the label they want to label (This then gets object id 0 for example), then i let them choose on which frame they want to click on the object, maybe let them click on multiple obejcts, with the live preview of the mask, and then i track that object through the video. This may take a few seconds but after completing i show the masks for that object on all frames.

Now if they want to switch the Label or make a new one, i store all clicked points and masks and reset the predictor. Sam2 still has all the images but is now ready to track the next object. (This new object gets id 1 and so on).
When the user wants to correct or change the first label, i store the current masks and points, reset the predictor, and load the points from the first label.
These points i give to sam directly, so that the last points they selected will also get considered.

The Tracking this way is faster, and you dont have to calculate for every label, just because you changed an unrelated one

**Comment by 2nzi** [comment]
@kyzira, thanks again for the helpful explanations!

I had a follow-up question, more related to training data preparation:
When tracking multiple objects (e.g., players in sports videos), do you create a separate video and mask per object, or do you include all objects in a single video with their own masks (e.g., different object IDs per frame)?

In other words, in your workflow, is it better to isolate each object and generate one training sample per object, or to annotate multiple objects within the same frames and track them all together?

**Comment by kyzira** [comment]
I have multiple objects per frame, the thing it gets better at, are the shapes. In the binary masks, i just use white, even if it is different objects

**Comment by 2nzi** [comment]
@kyzira That makes sense for simpler cases, but I’m a bit skeptical about this approach in my context — with sports videos, and when players are close together or occlude each other, the binary mask would just become one big block of white, making it hard to distinguish individual shapes.

Did you run into issues like that? Or did you find that the model still learned to separate the objects well despite that?

Another question came to mind:
If an object goes out of the frame for a moment, do you save a "black" mask along with the corresponding JPEG image in that case?

If i choose to export  1 image per second, i should name them 00000, 00001 and not the frame associate une the video like with 25 fps, 00025 for the first second image ?
